router/core/websocket.go

```diff
@@ -158,76 +128,215 @@ func (c *wsConnectionWrapper) Close() error {
 		}
-		r = validatedReq
-
-		upgrader := websocket.Upgrader{
-			HandshakeTimeout: 5 * time.Second,
-			// TODO: WriteBufferPool,
-			EnableCompression: true,
-			Subprotocols:      wsproto.Subprotocols(),
-			CheckOrigin: func(_ *http.Request) bool {
-				// Allow any origin to subscribe via WS
+		http.Error(w, http.StatusText(statusCode), statusCode)
+		return
+	}
+	r = validatedReq
+	var (
+		subProtocol string
+	)
+	upgrader := ws.HTTPUpgrader{
+		Timeout: time.Second * 5,
+		Protocol: func(s string) bool {
+			if wsproto.IsSupportedSubprotocol(s) {
+				subProtocol = s
 				return true
-			},
-		}
-		c, err := upgrader.Upgrade(w, r, nil)
+			}
+			return false
+		},
+	}
+	c, rw, _, err := upgrader.Upgrade(r, w)
+	if err != nil {
+		h.logger.Warn("websocket upgrade", zap.Error(err))
+		return
+	}
+	conn := newWSConnectionWrapper(h.ctx, c, rw)
+	protocol, err := wsproto.NewProtocol(subProtocol, conn)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		_ = c.Close()
+		return
+	}
+	handler := NewWebsocketConnectionHandler(h.ctx, WebSocketConnectionHandlerOptions{
+		Parser:         h.parser,
+		Planner:        h.planner,
+		GraphQLHandler: h.graphqlHandler,
+		Metrics:        h.metrics,
+		ResponseWriter: w,
+		Request:        r,
+		Connection:     conn,
+		Protocol:       protocol,
+		Logger:         h.logger,
+		Stats:          h.stats,
+		ConnectionID:   h.connectionIDs.Inc(),
+	})
+	err = handler.Initialize()
+	if err != nil {
+		h.logger.Warn("initializing websocket connection", zap.Error(err))
+		_ = c.Close()
+		return
+	}
+	// handle the first message in the main goroutine, which is faster than polling
+	// then add the connection to the epoll to save resources
+	err = h.HandleMessage(handler, h.readTimeout)
+	if err != nil && !h.isReadTimeout(err) {
+		h.logger.Warn("handling websocket connection", zap.Error(err))
+		_ = c.Close()
+		return
+	}
+	if h.epoll != nil {
+		err = h.addConnection(c, handler)
 		if err != nil {
-			// Upgrade() sends an error response already, just log the error
-			h.logger.Warn("upgrading websocket", zap.Error(err))
+			h.logger.Warn("adding connection to epoll", zap.Error(err))
+		}
+		return
+	}
+	go h.handleConnectionSync(handler)
+}
+
+func (h *WebsocketHandler) handleConnectionSync(handler *WebSocketConnectionHandler) {
+	h.stats.ConnectionsInc()
+	defer h.stats.ConnectionsDec()
+	serverDone := h.ctx.Done()
+	defer handler.Close()
+	for {
+		select {
+		case <-serverDone:
 			return
+		default:
+			err := h.HandleMessage(handler, h.readTimeout)
+			if err != nil {
+				if h.isReadTimeout(err) {
+					continue
+				}
+				if errors.Is(err, errClientTerminatedConnection) {
+					return
+				}
+				h.logger.Warn("handling websocket connection", zap.Error(err))
+				return
+			}
 		}
-		conn := newWSConnectionWrapper(h.ctx, c)
-		protocol, err := wsproto.NewProtocol(c.Subprotocol(), conn)
+	}
+}
+
+func (h *WebsocketHandler) isReadTimeout(err error) bool {
```

Comment:
I'd make it a pure function. No data is shared. Maybe we can even reuse it somewhere else.

Reply:
yes